nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/excel/RowIterator.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.excel;
+
+import com.github.pjfanning.xlsx.StreamingReader;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.ss.usermodel.Workbook;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RowIterator implements Iterator<Row>, Closeable {
+    private final Workbook workbook;
+    private final Iterator<Sheet> sheets;
+    private final Map<String, Boolean> requiredSheets;
+    private final int firstRow;
+    private final ComponentLog logger;
+    private Sheet currentSheet;
+    private Iterator<Row> currentRows;
+    private Row currentRow;
+
+    public RowIterator(InputStream in, List<String> requiredSheets, int firstRow, ComponentLog logger) {
+        this.workbook = StreamingReader.builder()
+                .rowCacheSize(100)
+                .bufferSize(4096)
+                .open(in);
+        this.sheets = this.workbook.iterator();
+        this.requiredSheets = requiredSheets != null ? requiredSheets.stream()
+                .collect(Collectors.toMap(key -> key, value -> Boolean.FALSE)) : new HashMap<>();
+        this.firstRow = firstRow;
+        this.logger = logger;
+    }
+
+    @Override
+    public boolean hasNext() {
+        setCurrent();
```

Comment:
This approach requires calling `hasNext()` before calling `next()`, which may be conventional, but should not be required. Instead, the next row should be set after calling `next()` because that method should move the iterator forward.

Reply:
Just trying to understand. With that refactor then one could never call hasNext since the instant variable currentRow is null.

Reply:
It should be possible to call `hasNext()` in any order. Taking a closer look, it seems like the current implementation of `setCurrent` needs further refinement because it sets `currentSheet` to `null` on every invocation. One general approach could be to set the initial sheet and row in the constructor.

In general, it should be possible to call `hasNext()` and unlimited number of times without advancing the iterator, and it calling `next()` should advance the iterator forward. When all sheets and rows have been read, `next()` should throw an `IllegalStateException` if called again.

Reply:
It looks like I could take out lines
```
currentSheet = null;
currentRows = null;
```
from setCurrent without any issues. Since the RowIterator is for internal use, could we just implement the next method and not the hasNext method? It seem there would be considerable amount of change needed just to have the hasNext method work.

Reply:
Although it may not be the most intuitive approach, having `hasNext()` throw an `UnsupportedOperationException` and changing callers to just use `next()` could work, although it raises a question of logic handling for the calling class. Another option could be folding the functionality into the calling class.

I think it would be cleaner in the end to rework the RowIterator, in addition to removing the `public` visibility from the class. I could take a look at reworking RowIterator and pushing the changes to this branch if you think the changes are too involved.

Reply:
I am reworking it

Reply:
I simplified the RowIterator by doing some up front processing in the constructor to determine which sheets need to be processed. That allowed for simplifying later logic to be uniform just to act on the the sheets needed. I hope its okay that I have a call to `setCurrent()` in the constructor in order to initialize the `currentRow` to allow hasNext to be called initially.